src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,320 @@
+
+    // if the first character is whitespace, it's a beginning of a new word
+    if (isspace(firstLetter)) {
+      ++subword.begin;
+      wordsFeatures.push_back({1, subwordScore, subwordScore});
+      wordByteRanges.push_back(subword);
+    } else {
+      // update last word byte range and word features
+
+      ByteRange& currentWord = wordByteRanges.back();
+      WordFeatures& currentWordFeatures = wordsFeatures.back();
+
+      currentWord.end = subword.end;
+      ++currentWordFeatures.numSubWords;
+      // incremental mean
+      currentWordFeatures.meanScore += (subwordScore - currentWordFeatures.meanScore) / currentWordFeatures.numSubWords;
+
+      if (currentWordFeatures.minScore > subwordScore) {
+        currentWordFeatures.minScore = subwordScore;
+      }
+    }
+  }
+
+  const float overallMean = sequence / subwordIdx;
+
+  return {wordByteRanges, wordsFeatures, overallMean};
+}
+
+QualityEstimator::Matrix QualityEstimator::convertToMatrix(const std::vector<WordFeatures>& wordsFeatures,
+                                                           const float overallMean) {
+  const size_t numFeatures = 4;  // numSubWords, meanScore, minScore and overallMean
+
+  Matrix matrix(wordsFeatures.size(), numFeatures);
+
+  size_t i = 0;
+
+  for (const WordFeatures& features : wordsFeatures) {
+    matrix.data[i++] = features.meanScore;
+    matrix.data[i++] = features.minScore;
+    matrix.data[i++] = features.numSubWords;
+    matrix.data[i++] = overallMean;
+  }
+
+  return matrix;
+}
+
+std::vector<float> QualityEstimator::LogisticRegressor::vectorResult(const IntgemmMatrix& features) const {
+  const Matrix modelScores = features * coefficients;
+
+  std::vector<float> scores(features.rows);
+
+  for (int i = 0; i < modelScores.rows; ++i) {
+    scores[i] = modelScores.at(i, 0) + intercept;
+  }
+
+  return scores;
+}
+
+QualityEstimator::IntgemmMatrix QualityEstimator::LogisticRegressor::transformFeatures(const Matrix& features) const {
+  QualityEstimator::IntgemmMatrix resultFeatures(features.rows, features.cols, intgemm::Int16::tile_info.a_rows,
+                                                 intgemm::Int16::tile_info.a_cols);
+
+  const auto getStds = [](const auto stds) { return stds != 0.0 ? stds : 1.0f; };
+
+  for (int i = 0; i < features.rows; ++i) {
+    for (int j = 0; j < features.cols; ++j) {
+      resultFeatures.at(i, j) = (features.at(i, j) - scale.means[j]) / getStds(scale.stds[j]);
+    }
+  }
+
+  return resultFeatures;
+}
+
+void QualityEstimator::LogisticRegressor::resultToProbabilities(std::vector<float>& linearPredictedValues) const {
+  for (float& value : linearPredictedValues) {
+    value = 1 / (1 + std::exp(-value));
+  }
+}
+
+std::vector<float> QualityEstimator::LogisticRegressor::predict(const Matrix& features) const {
+  std::vector<float> scores = vectorResult(transformFeatures(features));
+  resultToProbabilities(scores);
+  return scores;
```

Comment:
> Only benefit I see here is that what's happening is "described via function names

We decided to separate into different functions in order to keep the code with responsibilities small and separated, as suggested by Clean Code book so this was indeed the intention behind this change. 

IMHO (and by the book), it is always better to use function names instead of comments.

Reply:
We removed all methods mentioned while keeping the predict one in LR. Commit eb3af9f